### PR TITLE
✨  Windows Adapter (Part 2) - Windows component identifier

### PR DESCRIPTION
**Pull request by acabarbaye**
### Description

On Windows, some related USB interfaces (e.g. CDC/Mass storage) may have different UID depending on how they were defined. Some may have an instance ID, some may have a serial number. As they are related, they must be considered as the same USB device and their data must be aggregated.

### Test Coverage

- [x]  This change is covered by existing or additional automated tests.
- [ ]  Manual testing has been performed (and evidence provided) as automated testing was not feasible.
- [ ]  Additional tests are not required for this change (e.g. documentation update).

### CODE REVIEW

File: mbed_devices/_internal/windows/windows_identifier.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            or self.UID == other.serial_number
+        ):
+            return True
+        # Due to the complexity of determining the UID on Windows,
+        # we can assume that we are dealing with the same ID if IDs are subsets of each other.
+        return str(self.UID).startswith(str(other.UID)) or str(other.UID).startswith(str(self.UID))
+
+    @property
+    def actual_serial_number(self) -> str:
+        """Determines the actual serial number of a device from the different components."""
```
Comment (muchzill4) [inline mbed_devices/_internal/windows/windows_identifier.py]: Could you add more documentation/examples here? I think readers of this might not know what "instance ID" is. Also worth explaining what "different components" are.

File: mbed_devices/_internal/windows/windows_identifier.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        if not is_undefined_value(self.UID) and INSTANCE_ID_CHARACTER not in str(self.UID):
+            # the UID is not an instance ID and should therefore be the serial number.
+            return self.UID
+        return (
+            self.UID
+            if is_undefined_value(self.serial_number) or INSTANCE_ID_CHARACTER in str(self.serial_number)
+            else str(self.serial_number)
```
Comment (muchzill4) [inline mbed_devices/_internal/windows/windows_identifier.py]: Is it intentional that this can return `str(None)`?

File: mbed_devices/_internal/windows/windows_identifier.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        ):
+            return True
+        # Due to the complexity of determining the UID on Windows,
+        # we can assume that we are dealing with the same ID if IDs are subsets of each other.
+        return str(self.UID).startswith(str(other.UID)) or str(other.UID).startswith(str(self.UID))
+
+    @property
+    def actual_serial_number(self) -> str:
+        """Determines the actual serial number of a device from the different components."""
+        if not is_undefined_value(self.UID) and INSTANCE_ID_CHARACTER not in str(self.UID):
```
Comment (muchzill4) [inline mbed_devices/_internal/windows/windows_identifier.py]: `self.UID` is already a `str`? No need to convert it.

File: mbed_devices/_internal/windows/windows_identifier.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            # the UID is not an instance ID and should therefore be the serial number.
+            return self.UID
+        return (
+            self.UID
+            if is_undefined_value(self.serial_number) or INSTANCE_ID_CHARACTER in str(self.serial_number)
```
Comment (muchzill4) [inline mbed_devices/_internal/windows/windows_identifier.py]: Maybe it's worth extracting
`is_undefined_value(value) or INSTANCE_ID_CHARACTER in str(value)`
to a function outside of this class?

File: mbed_devices/_internal/windows/windows_identifier.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        if INSTANCE_ID_CHARACTER in str(self.UID):
+            return self.UID
+        return self.UID if is_undefined_value(self.serial_number) else str(self.serial_number)
+
+    def contains_genuine_serial_number(self) -> bool:
+        """Contains a genuine serial number and not an instance ID."""
+        serial_number = self.actual_serial_number
+        return bool(serial_number and INSTANCE_ID_CHARACTER not in str(serial_number))
+
+    def __hash__(self) -> int:
```
Comment (muchzill4) [inline mbed_devices/_internal/windows/windows_identifier.py]: This breaks the pythons data model - https://docs.python.org/3/reference/datamodel.html#object.__hash__ 
> The only required property is that objects which compare equal have the same hash value

File: mbed_devices/_internal/windows/windows_identifier.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+            # the UID is not an instance ID and should therefore be the serial number.
+            return self.UID
+        return (
+            self.UID
+            if is_undefined_value(self.serial_number) or INSTANCE_ID_CHARACTER in str(self.serial_number)
+            else str(self.serial_number)
+        )
+
+    @property
+    def instance_id(self) -> str:
```
Comment (muchzill4) [inline mbed_devices/_internal/windows/windows_identifier.py]: What's the difference between this and `actual_serial_number`? It seems to have similar characteristics.

File: mbed_devices/_internal/windows/windows_identifier.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
+"""Defines a Windows identifier."""
+from typing import NamedTuple, Any, Optional
+
+from mbed_devices._internal.windows.component_descriptor_utils import is_undefined_value
+
+INSTANCE_ID_CHARACTER = "&"
+
+
+class WindowsUID(NamedTuple):
```
Comment (rwalton-arm) [inline mbed_devices/_internal/windows/windows_identifier.py]: I think we should drop the `NamedTuple` inheritance if we're implementing `__eq__` and `__hash__` anyway, here is a nice breakdown of why we should probably avoid inheriting `NamedTuple` https://www.attrs.org/en/stable/why.html#namedtuples

File: mbed_devices/_internal/windows/windows_identifier.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+            or self.serial_number == other.UID
+            or self.UID == other.serial_number
+        ):
+            return True
+        # Due to the complexity of determining the UID on Windows,
+        # we can assume that we are dealing with the same ID if IDs are subsets of each other.
+        return str(self.UID).startswith(str(other.UID)) or str(other.UID).startswith(str(self.UID))
+
+    @property
+    def actual_serial_number(self) -> str:
```
Comment (rwalton-arm) [inline mbed_devices/_internal/windows/windows_identifier.py]: I'm not sure about the naming here, the `actual_` prefix seems redundant. Could we change it to just `serial_number`?

File: mbed_devices/_internal/windows/windows_identifier.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
+"""Defines a Windows identifier."""
+from typing import NamedTuple, Any, Optional
+
+from mbed_devices._internal.windows.component_descriptor_utils import is_undefined_value
+
+INSTANCE_ID_CHARACTER = "&"
+
+
+class WindowsUID(NamedTuple):
+    """Definition of a Windows Universal ID."""
```
Comment (rwalton-arm) [inline mbed_devices/_internal/windows/windows_identifier.py]: I'm not sure what a "Windows Universal ID" is. Might be worth adding a description or link to existing docs here.

File: mbed_devices/_internal/windows/windows_identifier.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    raw_uid: str  # The String from which the UID was derived. Raw form as it can be a
+    # combination of other elements e.g. 8&3432E179&0&000
+    serial_number: Optional[str]  # a serial number
+
+    def __eq__(self, other: Any) -> bool:
+        """Defines the equality checker."""
+        if not other or not isinstance(other, WindowsUID):
+            return False
+        if (
+            self.UID == other.UID
```
Comment (rwalton-arm) [inline mbed_devices/_internal/windows/windows_identifier.py]: If `UID` is always a subset of `raw_uid` there's probably not much point adding it to the equality check as we're already checking `raw_uid`.

File: mbed_devices/_internal/windows/windows_identifier.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+INSTANCE_ID_CHARACTER = "&"
+
+
+class WindowsUID(NamedTuple):
+    """Definition of a Windows Universal ID."""
+
+    UID: str  # Processed string which may be the closest to the Universal ID.
+    raw_uid: str  # The String from which the UID was derived. Raw form as it can be a
+    # combination of other elements e.g. 8&3432E179&0&000
+    serial_number: Optional[str]  # a serial number
```
Comment (rwalton-arm) [inline mbed_devices/_internal/windows/windows_identifier.py]: What's `serial_number` and how does it differ from `UID`? We seem to check they are equal to each other in the `__eq__` method.

File: tests/_internal/windows/test_windows_identifier.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        uid2 = WindowsUID(UID="uid12&223", raw_uid="djfds;fj", serial_number=None)
+        self.assertEqual(uid2.actual_serial_number, uid2.UID)
+        self.assertNotEqual(uid2.actual_serial_number, uid2.serial_number)
+        self.assertFalse(uid2.contains_genuine_serial_number())
+        uid3 = WindowsUID(UID="uid12&223", raw_uid="djfds;fj", serial_number="12345679")
+        self.assertNotEqual(uid3.actual_serial_number, uid3.UID)
+        self.assertEqual(uid3.actual_serial_number, uid3.serial_number)
+        self.assertTrue(uid3.contains_genuine_serial_number())
+
+    def test_isntanceid(self):
```
Comment (rwalton-arm) [inline tests/_internal/windows/test_windows_identifier.py]:
```suggestion
    def test_instanceid(self):
```

File: mbed_devices/_internal/windows/windows_identifier.py [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+        """Defines the equality checker."""
+        if not other or not isinstance(other, WindowsUID):
+            return False
+        if (
+            self.UID == other.UID
+            or self.raw_uid == other.raw_uid
+            or self.serial_number == other.serial_number
+            or self.serial_number == other.UID
+            or self.UID == other.serial_number
+        ):
```
Comment (rwalton-arm) [inline mbed_devices/_internal/windows/windows_identifier.py]: The logic of this `__eq__` method is hard to follow. I think it needs a more descriptive docstring.

File: tests/_internal/windows/test_windows_identifier.py (hunk 12 above)
Comment (acabarbaye) [reply tests/_internal/windows/test_windows_identifier.py]: 28301b80044ce0a6df8120eed868e2979d6718dd

File: mbed_devices/_internal/windows/windows_identifier.py (hunk 13 above)
Comment (acabarbaye) [reply mbed_devices/_internal/windows/windows_identifier.py]: 28301b80044ce0a6df8120eed868e2979d6718dd

File: mbed_devices/_internal/windows/windows_identifier.py (hunk 11 above)
Comment (acabarbaye) [reply mbed_devices/_internal/windows/windows_identifier.py]: tried to comment more

File: mbed_devices/_internal/windows/windows_identifier.py (hunk 10 above)
Comment (acabarbaye) [reply mbed_devices/_internal/windows/windows_identifier.py]: UID is not always a subset. it can be a lowercase subset.

File: mbed_devices/_internal/windows/windows_identifier.py (hunk 9 above)
Comment (acabarbaye) [reply mbed_devices/_internal/windows/windows_identifier.py]: tried to comment more about this 28301b80044ce0a6df8120eed868e2979d6718dd

File: mbed_devices/_internal/windows/windows_identifier.py [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+
+from mbed_devices._internal.windows.component_descriptor_utils import is_undefined_value
+
+INSTANCE_ID_CHARACTER = "&"
+
+
+class WindowsUID(NamedTuple):
+    """Definition of a Windows Universal ID."""
+
+    UID: str  # Processed string which may be the closest to the Universal ID.
```
[the remark this answers is not in the document]
Comment (acabarbaye) [reply mbed_devices/_internal/windows/windows_identifier.py]: changed it

File: mbed_devices/_internal/windows/windows_identifier.py (hunk 1 above)
Comment (acabarbaye) [reply mbed_devices/_internal/windows/windows_identifier.py]: Hopefully I have been more clear

File: mbed_devices/_internal/windows/windows_identifier.py (hunk 7 above)
Comment (acabarbaye) [reply mbed_devices/_internal/windows/windows_identifier.py]: maybe we should consider the use of `attrs` in the future but I would rather discuss about this as part of a more general decision

File: mbed_devices/_internal/windows/windows_identifier.py (hunk 8 above)
Comment (acabarbaye) [reply mbed_devices/_internal/windows/windows_identifier.py]: I agree `actual` was not the best descriptive adjective and I changed it for that matter

File: mbed_devices/_internal/windows/windows_identifier.py (hunk 11 above)
Comment (acabarbaye) [reply mbed_devices/_internal/windows/windows_identifier.py]: e3b58499695e9796bf1b0ed63d97dd330c1692a8

File: mbed_devices/_internal/windows/windows_identifier.py (hunk 10 above)
Comment (acabarbaye) [reply mbed_devices/_internal/windows/windows_identifier.py]: e3b58499695e9796bf1b0ed63d97dd330c1692a8

File: mbed_devices/_internal/windows/windows_identifier.py (hunk 8 above)
Comment (acabarbaye) [reply mbed_devices/_internal/windows/windows_identifier.py]: e3b58499695e9796bf1b0ed63d97dd330c1692a8

File: mbed_devices/_internal/windows/windows_identifier.py (hunk 6 above)
Comment (acabarbaye) [reply mbed_devices/_internal/windows/windows_identifier.py]: e3b58499695e9796bf1b0ed63d97dd330c1692a8

File: mbed_devices/_internal/windows/windows_identifier.py (hunk 5 above)
Comment (acabarbaye) [reply mbed_devices/_internal/windows/windows_identifier.py]: e3b58499695e9796bf1b0ed63d97dd330c1692a8

File: mbed_devices/_internal/windows/windows_identifier.py (hunk 4 above)
Comment (acabarbaye) [reply mbed_devices/_internal/windows/windows_identifier.py]: e3b58499695e9796bf1b0ed63d97dd330c1692a8

File: tests/_internal/windows/test_windows_identifier.py [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+        uid4 = WindowsUID(UID="uid4", raw_uid="/uid4&004", serial_number=uid1.serial_number)
+        uid5 = WindowsUID(UID="uid1&0114", raw_uid="/uid1&0114", serial_number=None)
+        uid6 = WindowsUID(UID="uid6454", raw_uid="/uid6454&006", serial_number="uid1")
+        self.assertNotEqual(uid1, None)
+        self.assertNotEqual(uid1, "")
+        self.assertNotEqual(uid1, uid2)
+        self.assertEqual(uid1, uid1)
+        self.assertTrue(uid1 == uid1)
+        self.assertEqual(uid1, uid3)
+        self.assertTrue(uid1 == uid3)
```
[the remark this answers is not in the document]
Comment (acabarbaye) [reply tests/_internal/windows/test_windows_identifier.py]: It should not be different but I am just ensuring that this is the case

File: mbed_devices/_internal/windows/windows_identifier.py (hunk 7 above)
Comment (rwalton-arm) [reply mbed_devices/_internal/windows/windows_identifier.py]: I'm not sure if we should use `attrs`. However, I do think for the reasons mentioned in the link, and due to the fact we define our own `__eq__`, `__repr__` etc, we shouldn't inherit from `NamedTuple`.

File: tests/_internal/windows/test_windows_identifier.py [hunk 16]
```diff
@@ -0,0 +1,83 @@
+import unittest
```
Comment (madchutney) [inline tests/_internal/windows/test_windows_identifier.py]: It would help me to see some comments around the tests as they define the behaviour.

File: mbed_devices/_internal/windows/windows_identifier.py [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+    serial_number: Optional[str]  # a serial number.
+    # A string which can be the serial number of the device defined by
+    # the underlying bus which may uniquely define the device.
+    # e.g. 345240562. However, for some devices,
+    # the serial number may not be accessible or defined or may actually be an instance ID.
+
+    def __eq__(self, other: Any) -> bool:
+        """Defines the equality checker.
+
+        The `equal` method does not follow a typical data model equal logic
```
[the remark this answers is not in the document]
Comment (muchzill4) [reply mbed_devices/_internal/windows/windows_identifier.py]: I agree with that. Seems to me we're trying to fit something into a box that doesn't really fit the requirements. The fact that this is needed up the stack for _reliable_ dict operations (merging data) might indicate there's an inherent problem with the implementation. :)

File: tests/_internal/windows/test_windows_identifier.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (muchzill4) [reply tests/_internal/windows/test_windows_identifier.py]: I'd get rid of this. No need to ensure python works. If it doesn't then we're screwed anyway.
